### PR TITLE
Lexer / Parser / Pretty for Sorbet (the new syntax)

### PR DESCRIPTION
**Pull request by jystic**
The is the first step towards the new syntax which looks a lot more like the syntax presented in Amos and Ben's paper and in addition has whitespace sensitive case expressions and let blocks.

To get an idea of what it looks like I've converted the prelude and put it in a new file `prelude-new.icicle`.

I started by documenting the existing grammar (in `icicle-v0.bnf`) and used that as a basis for the sorbet grammar (in `sorbet.bnf`).

The concrete syntax types match the grammar exactly 1:1 so you can read them as the grammar itself. This makes the parser really easy, it's just applicative over each constructors. The plan is to have an elaboration phase which converts the concrete syntax to the abstract syntax (i.e. the current icicle source AST).

/cc @amosr @tranma @markhibberd

### CODE REVIEW

File: doc/internals/icicle-v0.bnf [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+                |  <digit>
+                |  _
+                |  $
+                |  '
+<var-tails>    ::= <var-tails> <var-tail>
+                |  ε
+
+<con-id>       ::= <con-head> <con-tails>
+<con-head>     ::= <upper>
+                |  _
```
Comment (amosr) [inline doc/internals/icicle-v0.bnf]: I don't think constructors can start with underscore, and I guess single "_" case is just a var

Comment (jystic) [inline doc/internals/icicle-v0.bnf]: This is the v0 (current) syntax where I think this is actually the case, the new syntax is in `doc/internals/sorbet.bnf`

File: src/Icicle/Sorbet/Concrete/Annotation.hs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+  , annotOfConId
+  ) where
+
+import           Icicle.Sorbet.Concrete.Syntax
+
+
+annotOfRepl :: Repl a -> a
+annotOfRepl = \case
+  From a _ _ ->
+    a
```
Comment (amosr) [inline src/Icicle/Sorbet/Concrete/Annotation.hs]: are annotations comonads?

Comment (jystic) [inline src/Icicle/Sorbet/Concrete/Annotation.hs]: Discussed IRL, looks like they are!

File: doc/internals/sorbet.bnf [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+                    |  of
+                    |  if
+                    |  then
+                    |  else
+                    |  for
+                    |  fold
+                    |  fold1
+                    |  window
+                    |  group
+                    |  fold
```
Comment (HuwCampbell) [inline doc/internals/sorbet.bnf]: fold is mentioned twice

**Comment by amosr** [comment]
🍧 so good

File: doc/internals/sorbet.bnf (hunk 3 above)
Comment (jystic) [inline doc/internals/sorbet.bnf]: ta, fixed

**Pull request merged**